### PR TITLE
Reschedule and cancel

### PR DESCRIPTION
**Pull request by tedsteen**
Suggestion for #4

### CODE REVIEW

File: src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java [hunk 1]
```diff
@@ -32,12 +32,11 @@
 
 import static com.github.kagkarlsson.scheduler.ExecutorUtils.defaultThreadFactoryWithPrefix;
 
-public class Scheduler implements SchedulerClient {
+public class Scheduler extends SchedulerClient.StandardSchedulerClient {
```
Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java]: What is the reason SchedulerClient interface was changed for extends here? The idea behind SchedulerClient is that you are able to "downgrade" the full Scheduler object to only a SchedulerClient, not exposing everything

**Comment by kagkarlsson** [comment]
Made a small comment. Will have a more thorough look later

**Comment by kagkarlsson** [comment]
Except for my comment, looks good on first glance

File: src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java (hunk 1 above)
Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java]: Because the implementation for was the same for both of them (`Scheduler` and `StandardSchedulerClient`).

Given the fact that `StandardSchedulerClient` implements `SchedulerClient` you will still be able to "downgrade" the full Scheduler object to only a SchedulerClient.

My reasoning was that if `Scheduler` extends `StandardSchedulerClient`:
* they will share the implementation so we don't repeat ourselves
* No new functionality will be exposed on `Scheduler` as `StandardSchedulerClient` only implements `SchedulerClient`

WDYT?

Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java]: Ah, you are correct, and I agree with your reasoning. For some reason, that extends still makes me itch :). Another option is implementing the interface, but delegating to an internal `StandardSchedulerClient`. I know it is pretty cosmetic though..

**Comment by kagkarlsson** [comment]
I see now that there is a potential issue with this approach for cancel/reschedule. If the execution is executed by the scheduler after the `getExecution` but before the `cancel/reschedule`, the call will fail due to the version number being increased (the optimistic locking db-scheduler uses heavily). Also, modifying it while it is "locked" for execution could lead to some unpredictable result, so we should probably only allow the cancel/reschedule if it is not currently executing.

Not sure if there is a way around having custom reschedule and cancel methods.. enforcing `picked = false`. Maybe if we check for `picked=false` after the `getExecution`, and just let it fail if it has been picked for execution in the mean time. The only thing is that currently, the exception message is "Expected one execution to be updated, but updated X. Indicates a bug.", which is kinda misleading for this case that might actually happen. 

Sorry, this was kinda train-of-thoughty :). I will try and make comments for the code lines.

File: src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 2]
```diff
@@ -27,9 +28,13 @@
 public interface SchedulerClient {
 	
 	void scheduleForExecution(Instant exeecutionTime, TaskInstance taskInstance);
```
Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: Maybe we should do some renaming here. 
How about:
* schedule
* reschedule
* cancel

And maybe a default method for the old scheduleForExecution, possibly deprecated?

File: src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 3]
```diff
@@ -59,7 +64,27 @@ public SchedulerClient build() {
...
+			} else {
+				throw new RuntimeException(String.format("Could not reschedule - no task with name '%' and id '%s' was found." , taskName, instanceId));
+			}
+		}
+
+		@Override
+		public void cancelSchedule(String taskName, String instanceId) {
+			Optional<Execution> execution = taskRepository.getExecution(taskName, instanceId);
+			if(execution.isPresent()) {
+				taskRepository.remove(execution.get());
```
Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: if `picked = true` here, throw exception about not allowed to cancel executions that are currently executing

File: src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 4]
```diff
@@ -59,7 +64,27 @@ public SchedulerClient build() {
...
 				TaskInstance taskInstance) {
 			taskRepository.createIfNotExists(new Execution(exeecutionTime, taskInstance));
-		} 
+		}
+
+		@Override
+		public void reschedule(String taskName, String instanceId, Instant newExecutionTime) {
+			Optional<Execution> execution = taskRepository.getExecution(taskName, instanceId);
+			if(execution.isPresent()) {
+				taskRepository.reschedule(execution.get(), newExecutionTime, null, null);
```
Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: See comment for cancel

File: src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java (hunk 1 above)
Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java]: I see. Which one would you prefer?
* delegating to internal `StandardSchedulerClient`
* make an `AbstractSchedulerClient` and extend from that in both `StandardSchedulerClient` and `Scheduler`

File: src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java (hunk 2 above)
Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: Sounds like a plan!

File: src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java (hunk 3 above)
Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: 👍

File: src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java (hunk 1 above)
Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java]: I think I would prefer keeping the interface and delegate to internal `StandardSchedulerClient`

Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java]: I think I'd prefer that as well actually. I'll get crackin' :)

File: src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java [hunk 5]
```diff
@@ -25,11 +26,23 @@
...
+	 * This method has been renamed and will be removed in a future version.
+	 *
+	 * @deprecated use {@link #schedule(Instant, TaskInstance)} instead.
+	 */
+	@Deprecated
+	void scheduleForExecution(Instant executionTime, TaskInstance taskInstance);
+
+	void schedule(Instant executionTime, TaskInstance taskInstance);
+
+	void reschedule(String taskName, String instanceId, Instant newExecutionTime);
```
Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: Changes looking good. Starting to feel I lack a type here, a `TaskInstanceId` or something. Previously that was just a `TaskInstance`, but now a `TaskInstance` is something more, it can have data as well. I will think about it, maybe I will change it in a later version :)

Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: You're right. It seems to be quite intrusive so, as you say, might be better to do in another PR/version.

Comment (tedsteen) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: Btw, I did a small PR in my fork to try it out. If you like it I can merge and it'll be part of this PR.
https://github.com/tedsteen/db-scheduler/pull/1

Comment (kagkarlsson) [inline src/main/java/com/github/kagkarlsson/scheduler/SchedulerClient.java]: Nice! Doesn't look like it required too many changes to introduce the type. Might as well make it part of this PR.

Was thinking we should also change the order of parameters of the `schedule` method, for symmetry...  
```
void schedule(TaskInstance taskInstance, Instant executionTime);
```

**Comment by kagkarlsson** [comment]
Looks good, thanks for the PR! :)

Merging to master

**Comment by tedsteen** [comment]
Woho!

When do you think you could make a release so it could be picked up through maven?

**Comment by kagkarlsson** [comment]
Just deployed a 1.10-SNAPSHOT, but I'm assuming you mean release-version. I
think I just need to update a couple of things in the readme, then it
should be good to go. Hopefully within a couple of days.. depends on how
much time I can find :)

2017-05-02 21:30 GMT+02:00 Ted Steen <notifications@github.com>:

**Comment by tedsteen** [comment]
Actually, 1.10-SNAPSHOT will give me the master right?
In that case, there's no urgency at all from my end. I'll just use the snapshot.

**Comment by kagkarlsson** [comment]
Yeah, it's from master. Let me know if it is not available via maven


tir. 2. mai 2017 kl. 21.39 skrev Ted Steen <notifications@github.com>:
